src/IECoreMaya/SceneShapeSubSceneOverride.cpp

```diff
@@ -0,0 +1,1615 @@
+	IECore::ConstV2fVectorDataPtr uv = meshPrimitiveTriangulated->expandedVariableData<IECore::V2fVectorData>( "uv", IECoreScene::PrimitiveVariable::FaceVarying );
+
+	// We need topology information to render wireframes. When changing the
+	// data in `ensureFaceVaryingData`, we invalidate these indices, but the
+	// function allows passing a list of indices to be updated along with the
+	// data.
+	std::vector<int> wireframeIndicesReadable;
+	if( wireframeIndicesOriginal )
+	{
+		wireframeIndicesReadable = std::vector<int>( wireframeIndicesOriginal->readable() );
+	}
+
+ 	// Expand out positions and normals so that they line up with UVs
+	ensureFaceVaryingData( indicesTriangulated, p, n, wireframeIndicesReadable, linearInterpolation );
+
+	geometryData->positionData = p;
+	geometryData->normalData = n;
+	geometryData->uvData = uv;
+
+	// Prepare the index buffer for rendering a solid mesh
+	geometryData->indexData = indicesTriangulated;
+
+	IECore::IntVectorDataPtr wireframeIndexData( new IntVectorData() );
+	std::vector<int> &wireframeIndexReadable = wireframeIndexData->writable();
+
+	const IECore::IntVectorData *numVerticesPerFace = meshPrimitive->verticesPerFace();
+	if( numVerticesPerFace )
+	{
+		const std::vector<int> &numVerticesPerFaceReadable = numVerticesPerFace->readable();
+
+		int totalNumEdgeIndices = std::accumulate( numVerticesPerFaceReadable.begin(), numVerticesPerFaceReadable.end(), 0 ) * 2;
+		wireframeIndexReadable.reserve( totalNumEdgeIndices );
+
+		int offset = 0;
+		for( int numVerts : numVerticesPerFaceReadable )
+		{
+			for( int k = 0; k < numVerts; ++k )
+			{
+				wireframeIndexReadable.push_back( wireframeIndicesReadable[offset + k] );
+				wireframeIndexReadable.push_back( wireframeIndicesReadable[offset + (k+1)%numVerts] );
+			}
+			offset += numVerts;
+		}
+	}
+	geometryData->wireframeIndexData = wireframeIndexData;
+}
+
+GeometryDataPtr geometryGetter( const GeometryDataCacheGetterKey &key, size_t &cost )
+{
+	GeometryDataPtr geometryData( new GeometryData() );
+
+	if( !key.object )
+	{
+		fillBoundData( key.bounds, geometryData );
+		return geometryData;
+	}
+
+	switch( static_cast<IECoreScene::TypeId>( key.object->typeId() ) )
+	{
+	case IECoreScene::TypeId::MeshPrimitiveTypeId :
+		fillMeshData( key.object, key.descriptorList, geometryData ); break;
+	case IECoreScene::TypeId::PointsPrimitiveTypeId :
+		fillPointsData( key.object, geometryData ); break;
+	case IECoreScene::TypeId::CurvesPrimitiveTypeId :
+		fillCurvesData( key.object, geometryData ); break;
+	default :
+		// we should never get here as only the above types are supported, see `objectCanBeRendered`.
+		break;
+	}
+
+	// \todo compute cost from geometry data
+	cost = 1;
+
+	return geometryData;
+}
+
+// Cache for geometric data
+using GeometryDataCache = IECore::LRUCache<IECore::MurmurHash, GeometryDataPtr, IECore::LRUCachePolicy::Parallel, GeometryDataCacheGetterKey>;
+GeometryDataCache g_geometryDataCache(geometryGetter, 100);
```

Comment:
I addressed this in 48f2022. 

I wasn't sure about the defaultObjectPool() solution as that would mean that we'd have to expose the GeometryData implementation more formally, - deriving from Object and giving it a TypeId? Is that correct? Also, are we concerned about interfering/slowing down other systems that use that cache? I've used a separate limit for that one for now. 

Note for future reference: Maya's MRenderer class provides access to `GPUtotalMemorySize()` and `GPUUsedMemorySize()`. Might come in handy at some point.

Reply:
Yeah, I'm not sure really. From a user perspective, its a pity we have so many different caches... its really hard for a user to say "Cortex can use N mb in Maya" right now... Minimally we'll need to re-run your various scenes to gauge performance in the different production scenarios and determine if 500mb is a reasonable limit (at IE set the defaultObjectPool to 4gb for example, and some users increase it much higher). 500 is fine for the code, we just need to know how we'll set that env var here.

@johnhaddon any thoughts on whether to make a dedicated VP2 cache or try to hook into some other existing Cortex cache?

Reply:
Actually, I realise I don't even know why we're caching the GeometryData. That's an intermediate form isn't it? And Maya will have its own representation which is actually used for the drawing? If so, can we just cache the final form? Sorry if this is a dumb question...

Reply:
Not a dumb question, at all. 

It's an intermediate form, yes, and the final form are the `MRenderItem` objects that Maya uses for rendering. I initially played with caching just those guys, but that led to really messy code. The reason being that once we hand them off to Maya by putting them in the container Maya provides, Maya assumes ownership and takes the right to destroy them at any time (potentially once the GPU memory is filled up, and definitely also when we call `clear()` on the container). So, holding on to them is no good. My next test was then to always hand off copies to Maya, but the Maya API really isn't made for that and they really expect you to use the API in the way we do now. While you can copy MRenderItems for example, you cannot rename them and because the container indexes them by name, you can't put in a copy twice (for identical instances for example). All of that was really messy and I eventually converged to being closer to their examples which is what we're reviewing at the moment... I probably should have done that from the beginning, but maintaining the stash of objects to be rendered myself was tempting since we have the LRUCache and learning how Maya wants things done was a bit of a process.

As a side note, my hope with this now is that for the second iteration it will make sense to use the hash we get for the GeometryData as the name of the respective MRenderItems in the container. Looking up the MRenderItems and appending an instance matrix for every successful lookup should then give us instanced rendering and the performance improvements I saw from this earlier. That might me fairly low hanging fruit at this point - if the rendering works and geo isn't dropping out anymore.